codec-http2/src/main/java/io/netty/handler/codec/http2/ChildChannel.java

```diff
@@ -0,0 +1,275 @@
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import io.netty.channel.AbstractChannel;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelConfig;
+import io.netty.channel.ChannelMetadata;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.DefaultChannelConfig;
+import io.netty.channel.EventLoop;
+import io.netty.channel.SingleThreadEventLoop;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.InternalThreadLocalMap;
+import io.netty.util.internal.OneTimeTask;
+
+import java.net.SocketAddress;
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Child {@link Channel} of another channel.
+ */
+abstract class ChildChannel extends AbstractChannel {
+    private static Logger log = Logger.getLogger(ChildChannel.class.getName());
+    private static final ChannelMetadata METADATA = new ChannelMetadata(false);
+    private static final int MAX_READER_STACK_DEPTH = 8;
+
+    private final ChannelConfig config = new DefaultChannelConfig(this);
+    private final Queue<Object> inboundBuffer = new ArrayDeque<Object>();
+    private final Runnable readTask = new Runnable() {
+        @Override
+        public void run() {
+            ChannelPipeline pipeline = pipeline();
+            for (;;) {
+                Object m = inboundBuffer.poll();
+                if (m == null) {
+                    break;
+                }
+                pipeline.fireChannelRead(m);
+                channelReadOccurred(m);
+            }
+            pipeline.fireChannelReadComplete();
+        }
+    };
+    private final Runnable fireChildReadCompleteTask = new Runnable() {
+        @Override
+        public void run() {
+            if (readInProgress) {
+                readInProgress = false;
+                pipeline().fireChannelReadComplete();
+            }
+        }
+    };
+
+    private final Runnable shutdownHook = new Runnable() {
+        @Override
+        public void run() {
+            unsafe().close(unsafe().voidPromise());
+        }
+    };
+
+    private boolean closed;
+    private boolean readInProgress;
+
+    public ChildChannel(Channel parent) {
+        super(parent);
```

Comment:
"Manual" is the mechanism :-).

Right now I'm linking the stream state with the child channel state, so if the stream is closed I make sure to make the child channel inactive. The other thing I've hooked up (in my upcoming changes, because of this comment) is onStreamError will propagate the exception to the child channel.